src/liballoc/collections/binary_heap.rs

```diff
@@ -1166,6 +1166,19 @@ impl<T> ExactSizeIterator for IntoIter<T> {
 #[stable(feature = "fused", since = "1.26.0")]
 impl<T> FusedIterator for IntoIter<T> {}
 
+#[unstable(issue = "0", feature = "inplace_iteration")]
+unsafe impl<T> SourceIter for IntoIter<T> {
+    type Source = crate::vec::IntoIter<T>;
```

Comment:
This exposes internal implementation details in a way that it probably shouldn't.

Reply:
I am not sure what to do about that.

* remove it
* hide behind another, crate-private trait as layer of indirection over `binary_heap::IntoIter` and `vec::IntoIter` and specialize on that instead
* duplicate all the specialization
* leave as is, revisit later considering it's unstable

Reply:
> hide behind another, crate-private trait as layer of indirection over binary_heap::IntoIter and vec::IntoIter and specialize on that instead

That's usually what we do with specialization I think.

Reply:
The downside is that external crates (e.g. itertools) can't use this for specialization then

Reply:
It doesn't seem to work anyway

```
error[E0445]: crate-visible trait `vec::AsIntoIter<T>` in public interface
    --> src/liballoc/vec.rs:2708:5
     |
2708 |     type Source = impl AsIntoIter<T>;
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ can't leak crate-visible trait
...
2716 | pub(crate) trait AsIntoIter<T>: Iterator<Item=T> {
     | ---------- `vec::AsIntoIter<T>` declared as crate-visible

Reply:
But it turns out I can hide it behind an `impl Iterator` and specialization still sees through it. Is that good enough?

Reply:
The current test failures are caused by an [unimplemented! in librustdoc](https://github.com/rust-lang/rust/blob/a605441e049f0b6d5f7715b94b8ac4662fd7fcf6/src/librustdoc/clean/mod.rs#L1297) which is triggered by the *impl Trait* approach. Adding `#[doc(hidden)]` doesn't fix it, rustdoc still wants to process the item.

Would reverting to the concrete type and simply hiding it from documentation be acceptable? It's unstable anyway.

Reply:
How crucial is having this fast path for BinaryHeap specifically?

Reply:
Not very

Reply:
I have settled for a different way of applying the crate-private trait instead, the BinaryHeap internals should be properly hidden now.